### PR TITLE
ARO-1422: UpgradeConfig Policy 

### PR DESCRIPTION
**Pull request by edisonLcardenas**
### Which issue this PR addresses:

[ARO-1422: UpgradeConfig](https://issues.redhat.com/browse/ARO-1422)

### What this PR does / why we need it:
Block editing of UpgradeConfig by regular users. This will reduce amount of ticket we receive due to customers trying to do an unsupported action. We have a policy that will deny regulars users editing of UpgradeConfig.

### Test plan for issue:
* OPA Tests - **Passed**
* Gator Tests - **Passed**
* Real Cluster Verification - **Passed**
* Web Console Test - **Passed**

For detailed test result, please see my jira [ARO-1422: UpgradeConfig](https://issues.redhat.com/browse/ARO-1422).

### Is there any documentation that needs to be updated for this PR?
Yes, the README for guardrails. Included  info on how to create policies that needs to sync data into OPA using data.inventory object.

### CODE REVIEW

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/allow_system_user_delete_managed_upgrade_operator.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    kind: ConfigMap
+    version: v1
+  resource:
+    group: ""
+    resource: ConfigMap
+    version: v1
+  uid: 2ed5686f-74cd-41e2-96a2-afcf766fae51
+  userInfo:
+    uid: 1527f3d6-2e89-4b5f-a4bf-8953365c2016
+    username: system
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/allow_system_user_delete_managed_upgrade_operator.yaml:49]: lint fix here

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src_test.rego [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          "version": "v1"
+        },
+        "uid": "080407a7-d907-4c9a-8b17-67b637b97dce",
+        "userInfo": {
+          "uid": "0e4ced82-9535-4eaf-b65c-b091754cbd20",
+          "username": username
+        }
+      }
+    }
+  }
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src_test.rego:159]: lint fix :）

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/suite.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    assertions:
+    - violations: no
+  - name: delete-upgradeconfig-allowed-system-user
+    object: gator-test/allow_system_user_delete_managed_upgrade_operator.yaml
+    assertions:
+    - violations: no
+  - name: delete-upgradeconfig-allowed-regular-user
+    object: gator-test/not_allow_regular_user_delete_managed_upgrade_operator.yaml
+    assertions:
+    - violations: yes
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/suite.yaml:45]: also here

**Review by ArrisLee — APPROVED** [review]
LGTM over all, just some lint fix, and will be good if you can test the policy in a real dev cluster

**Comment by jhoreman** [comment]
Hey Edison. Can you ensure you sign/agree to the  Contributor License Agreement? Reach out on slack if you need assistance with this :)

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gatekeeper-config.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    - excludedNamespaces: ["kube-*","openshift-kube-*","openshift","openshift-etcd*","openshift-monitoring","default","gatekeeper-system","openshift-apiserver*","openshift-authentication*","openshift-logging","openshift-redhat-marketplace","openshift-operators","openshift-user-workload-monitoring","openshift-pipelines","openshift-marketplace","openshift-multus","openshift-network*","openshift-vsphere-*","openshift-config-*","openshift-console","openshift-service-ca*","openshift-azure*","openshift-cloud*","openshift-sdn"]
+      processes: ["*"]
+  sync:
+    syncOnly:
+      - group: ""
+        version: "v1"
+        kind: "ConfigMap"
+      - group: ""
+        version: "v1"
+        kind: "Namespace"
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gatekeeper-config.yaml:17]: NIT: lint fix here

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/regular_user_create_managed_upgrade_operator.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    kind: UpgradeConfig
+    version: upgrade.managed.openshift.io/v1alpha1
+  resource:
+    group: ""
+    resource: UpgradeConfig
+    version: upgrade.managed.openshift.io/v1alpha1
+  uid: e952ffa6-fd19-4d17-a7da-a98cb510c061
+  userInfo:
+    uid: eed44a0f-c0e7-469c-a489-9c3970988f2c
+    username: fake-k8s-regular-user
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/regular_user_create_managed_upgrade_operator.yaml:36]: linx fix

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/system_user_create_managed_upgrade_operator.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    kind: UpgradeConfig
+    version: upgrade.managed.openshift.io/v1alpha1
+  resource:
+    group: ""
+    resource: UpgradeConfig
+    version: upgrade.managed.openshift.io/v1alpha1
+  uid: e952ffa6-fd19-4d17-a7da-a98cb510c061
+  userInfo:
+    uid: eed44a0f-c0e7-469c-a489-9c3970988f2c
+    username: system:admin
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/gator-test/system_user_create_managed_upgrade_operator.yaml:36]: and here ;)

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src.rego [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	# ## has NO cloud.openshift.com entry in openshift-config/pull-secret Secret
+	# ## ALLOW EDITING
+
+	# ## If regular user and 
+	# ## HAS cloud.openshift.com entry (`source: OCM` indicates pull-secret exists) in openshift-config/pull-secret Secret
+	# ## NOT ALLOWED
+	config_data := data.inventory.namespace["openshift-managed-upgrade-operator"]["v1"]["ConfigMap"]["managed-upgrade-operator-config"]["data"]["config.yaml"]
+	regex.match("source: OCM", config_data)
+	msg := "Modifying the UpgradeConfig is not allowed for regular users. This includes attempting to create, edit, or delete the UpgradeConfig."
+}
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src.rego:25]: lint fix

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src_test.rego [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    output = {
+              "apiVersion": "v1",
+              "managed-upgrade-operator-config" : {
+                "data": {
+                    "config.yaml": "configManager:\n  source: LOCAL\n  \n  localConfigName: managed-upgrade-config\n\n watchInterval: 15\nmaintenance:\n  controlPlaneTime: 90\n  ignoredAlerts:\n\n   controlPlaneCriticals:\n    - ClusterOperatorDown\n    - ClusterOperatorDegraded\nupgradeType:\nARO\nupgradeWindow:\n  delayTrigger: 30\n  timeOut: 120\nnodeDrain:\n  timeOut:\n45\n  expectedNodeDrainTime: 8\nscale:\n  timeOut: 30\nhealthCheck:\n  ignoredCriticals:\n\n - PrometheusRuleFailures\n  - CannotRetrieveUpdates\n  - FluentdNodeDown\n\n ignoredNamespaces:\n  - openshift-logging\n  - openshift-redhat-marketplace\n\n - openshift-operators\n  - openshift-user-workload-monitoring\n  - openshift-pipelines\n\n - openshift-azure-logging\n"
+                }
+              },
+              "kind": "ConfigMap"
+    }
+}
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src_test.rego:106]: lint fix

File: pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src.rego [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+	not is_exempted_account(input.review)
+
+	# ## If regular user and
+	# ## has NO cloud.openshift.com entry in openshift-config/pull-secret Secret
+	# ## ALLOW EDITING
+
+	# ## If regular user and 
+	# ## HAS cloud.openshift.com entry (`source: OCM` indicates pull-secret exists) in openshift-config/pull-secret Secret
+	# ## NOT ALLOWED
+	config_data := data.inventory.namespace["openshift-managed-upgrade-operator"]["v1"]["ConfigMap"]["managed-upgrade-operator-config"]["data"]["config.yaml"]
```
Comment (ArrisLee) [inline pkg/operator/controllers/guardrails/policies/gktemplates-src/aro-deny-upgradeconfig/src.rego:22]: so my understanding here is, we are not able to retrieve the `config.yaml` data directly from `input review` object right?

**Pull request closed without merging**